### PR TITLE
[WIP] Lenses.jl rework

### PR DESCRIPTION
**Pull request by alfredclwong**
Opening a draft PR to discuss existing code using the review mechanism.

### CODE REVIEW

File: src/Optical/Lenses.jl [hunk 1]
```diff
@@ -2,123 +2,374 @@
...
+        frontsurfacereflectance = 0.0,
+        backsurfacereflectance = 0.0,
+        frontdecenter = (0, 0),
+        backdecenter = (0, 0),
+        interfacemode = ReflectOrTransmit
+    )
 
 Constructs a simple cylindrical lens with spherical front and back surfaces. The side walls of the lens are absorbing.
 """
-function SphericalLens(insidematerial::T, frontvertex::S, frontradius::S, backradius::S, thickness::S, semidiameter::S; lastmaterial::Q = OpticSim.GlassCat.Air, nextmaterial::R = OpticSim.GlassCat.Air, frontsurfacereflectance::S = zero(S), backsurfacereflectance::S = zero(S), frontdecenter::Tuple{S,S} = (zero(S), zero(S)), backdecenter::Tuple{S,S} = (zero(S), zero(S)), interfacemode = ReflectOrTransmit) where {R<:OpticSim.GlassCat.AbstractGlass,Q<:OpticSim.GlassCat.AbstractGlass,T<:OpticSim.GlassCat.AbstractGlass,S<:Real}
```
Comment (alfredclwong) [inline src/Optical/Lenses.jl:12]: Was there any reason for parameterising `AbstractGlass` types `R`, `Q`, `T`? This was done throughout the file.

File: src/Optical/Lenses.jl [hunk 2]
```diff
@@ -2,123 +2,376 @@
...
+    nsamples::Int = 17,
+    frontdecenter::Tuple{T,T} = (zero(T), zero(T)),
+    backdecenter::Tuple{T,T} = (zero(T), zero(T)),
+    interfacemode::InterfaceMode = ReflectOrTransmit
+) where {T<:Real}
+    @assert semidiameter > zero(T)
     @assert !isnan(frontradius)
 
-    frontdecenter_l = abs(frontdecenter[1]) > eps(S) && abs(frontdecenter[2]) > eps(S) ? sqrt(frontdecenter[1]^2 + frontdecenter[2]^2) : zero(S)
-    backdecenter_l = abs(backdecenter[1]) > eps(S) && abs(backdecenter[2]) > eps(S) ? sqrt(backdecenter[1]^2 + backdecenter[2]^2) : zero(S)
```
Comment (alfredclwong) [inline src/Optical/Lenses.jl:43]: Is it possible to do `x != zero(S)` instead of `abs(x) > eps(S)` here? Not sure if there's a particular use case of decentering that we are accounting for.

File: src/Optical/Lenses.jl [hunk 3]
```diff
@@ -2,123 +2,376 @@
...
+        zero(T)
+    )
+    backdecenter_l = (
+        abs(backdecenter[1]) > eps(T) && abs(backdecenter[2]) > eps(T) ?
+        sqrt(backdecenter[1]^2 + backdecenter[2]^2) :
+        zero(T)
+    )
+
+    # front lens
+    frontinterface = opticinterface(T, insidematerial, lastmaterial, frontsurfacereflectance, interfacemode)
```
Comment (alfredclwong) [inline src/Optical/Lenses.jl:200]: Other than interface, is there any reason to treat the front and back lenses separately? I'm reading and comparing both halves of the code to spot the differences now.

File: src/Optical/Lenses.jl [hunk 4]
```diff
@@ -2,123 +2,376 @@
...
+        lens_rear = leaf(
+            surf,
+            Transform{T}(zero(T), convert(T, π), zero(T), backdecenter[1], backdecenter[2], frontvertex - thickness)
+        )
     end
-    extra_front = frontradius >= zero(S) || isinf(frontradius) ? zero(S) : abs(frontradius) - sqrt(frontradius^2 - semidiameter^2)
-    extra_back = backradius >= zero(S) || isinf(backradius) ? zero(S) : abs(backradius) - sqrt(backradius^2 - semidiameter^2)
+
+    extra_front = (
+        frontradius >= zero(T) || isinf(frontradius) ?
```
Comment (alfredclwong) [inline src/Optical/Lenses.jl:322]: Does the first condition make the second condition redundant?

File: src/Optical/Lenses.jl (hunk 1 above)
Comment (friggog) [reply src/Optical/Lenses.jl:12]: Don't think so

File: src/Optical/Lenses.jl (hunk 2 above)
Comment (friggog) [reply src/Optical/Lenses.jl:43]: I think eps was used to allow for some numeric error as the decenter values are calculated, not sure if there is much practical impact to use zero instead - maybe add a test for this edge case?

File: src/Optical/Lenses.jl (hunk 3 above)
Comment (friggog) [reply src/Optical/Lenses.jl:200]: Front and back could have different reflectance - in the future they may have totally different interfaces due to different coatings being present on each side (very common)

Comment (alfredclwong) [reply src/Optical/Lenses.jl:200]: Vague wording from me here - I was trying to get an understanding of how much of the somewhat duplicated code is unique between the front/back lenses. My motivation for this is that I'll probably be splitting the lens code into

1. Various functions for creating a (front or back) lens surface of each surface type
2. A function for combining each of these surfaces into a CSG lens

I wanted to double-check that the only significant differences between front and back would be their interfaces and a z-translation. Is this correct?
